doc/design/dist/README.md

```diff
@@ -0,0 +1,78 @@
+# Distributed Training Design Doc
+
+## Objective
+
+We want Paddle to support training on a general-purpose cluster. The cluster runs Paddle, the Web server (e.g., Nginx), the log collector (e.g., fluentd), the distributed queue service (e.g., Kafka), the log joiner and other data processors written using Storm, Spark, and Hadoop MapReduce on the same cluster. As illustrated in the following graph:
+
+![general purpose cluster](src/arch.png)
+
+This poses new challenges for Paddle,
+
+- Paddle need to be tault tolerant.
+- Input training data can be online data from realtime logs, or batched data from distributed file system.
+- User needs a simple way to train model on cloud. Complexities such as job scheduling should be hidden from user.
+
+## Training Job
+
+A training job will be created once user asks Paddle cloud to train a model. The training job is made up of different processes that collabratively consume data input and produce a trained model. There are three kind of processes:
+
+- Master process
+- Trainer process
+- Parameter server process
+
+One training job will only have one master process, typicall multiple trainer and parameter server processes. Their relation is illustrated in the following graph:
+
+![process collabration](src/paddle-on-kubernetes-invited-blog-model-sharding.png)
+
+### Master Process
+
+Master process will:
+
+- keep a list of alive trainers and a list of alive parameter servers and do *health check*,
+  - if trainer is dead it will update task queue accordingly as mentioned in [task queue](#task-queue).
+  - if a parameter server is dead or a new parameter server joins, it will broacast this information to all trainers.
+- dispatches tasks to trainers. A *task* is a unit of data that a trainer needs to train on, and
+- keep track of training progress on the dataset with *task queue*. Typically training will iterate on the dataset for a full pass until it goes into next pass.
+
+#### Task Queue
+
+Master process have three task queues to track training progress as shown in the graph below:
```

Comment:
We need to define task somewhere, so could we start explaining task queues.

There are two ways to define tasks:

1. as in Hadoop MapReduce, a task is a sub-sequence of key-value pairs in a data file, or
1. as in Google MapReduce, a task is a file.